Multiple improvements to the inline assembly code

josephlr opened the pull request:
See the individual commits for more information. 

Amongst other things, this PR adds `external_asm` implementations for some functions that were missing it and documents those functions which still require `inline_asm`

Example Docs:

![docs](https://user-images.githubusercontent.com/5506060/117274004-83921c80-ae11-11eb-8e7e-a551c7fefbc8.png)

phil-opp commented on src/instructions/interrupts.rs:
```diff
@@ -152,19 +152,14 @@ pub fn int3() {
-    ($x:expr) => {{
-        compile_error!("software_interrupt not implemented for non-nightly");
+        #[cfg(not(feature = "inline_asm"))]
+        compile_error!("software_interrupt!() requires Nightly Rust");
```
We should probably mention that the `nightly` feature needs to be enabled (using a nightly compiler is not enough).

phil-opp commented on src/instructions/mod.rs:
```diff
@@ -53,14 +53,16 @@ pub fn bochs_breakpoint() {
     }
+
+    #[cfg(not(feature = "inline_asm"))]
+    compile_error!("read_rip() requires Nightly Rust");
```
Same as above:

> We should probably mention that the `nightly` feature needs to be enabled (using a nightly compiler is not enough).

phil-opp reviewed (approved):
Thanks a lot! Looks all good to me. Displaying the feature requirements on docs.rs is a great idea!

josephlr commented:
@phil-opp it looks like using `compile_error!()` in `read_rip()` doesn't do what we want (it breaks the whole build instead of just removing that function if `external_asm` is used). So I removed it, I also removed the similar usage from `software_interrupt!{}` so we are consistent.

Finally, I changed the docs to mention `nightly` and `inline_asm`, so they look like:


If this looks good to you, feel free to merge it.

phil-opp commented on src/registers/rflags.rs:
```diff
@@ -119,16 +119,13 @@ mod x86_64 {
     /// flags also used by Rust/LLVM can result in undefined behavior too.
     #[inline]
     pub unsafe fn write_raw(val: u64) {
+        // HACK: we mark this function as preserves_flags to prevent Rust from restoring
+        // saved flags after the "popf" below. See above note on safety.
```
Thanks for improving this comment, it's much clearer now!

phil-opp reviewed (approved):
Thanks a lot!

phil-opp commented:
Published as v0.14.2
